source/mir/ndslice/algorithm.d

```diff
@@ -403,46 +403,72 @@ unittest
 
 /++
 The call `eachUploPair!(fun)(matrix)`
-evaluates `fun` for each pair (`matrix[j, i]`, `matrix[i, j]`), i<=j.
+evaluates `fun` for each pair (`matrix[j, i]`, `matrix[i, j]`),
+for i <= j (default) or i < j (if includeDiagonal is false).
 
 Params:
     fun = A function.
+    includeDiagonal = true (default) if applying function to diagonal, 
+                      false otherwise.
 +/
-template eachUploPair(alias fun)
+template eachUploPair(alias fun, bool includeDiagonal = true)
 {
     import mir.functional: naryFun;
     static if (__traits(isSame, naryFun!fun, fun))
-    /++
-    Params:
-        matrix = Square matrix.
-    +/
-    auto eachUploPair(SliceKind kind, Iterator)(Slice!(kind, [2], Iterator) matrix)
-    in
-    {
-        assert(matrix.length!0 == matrix.length!1, "matrix must be square.");
-    }
-    body
     {
-        static if (kind == Contiguous)
+        /++
+        Params:
+            matrix = Square matrix.
+        +/
+        auto eachUploPair(SliceKind kind, Iterator)
+                                            (Slice!(kind, [2], Iterator) matrix)
+        in
         {
-            import mir.ndslice.topology: canonical;
-            .eachUploPair!fun(matrix.canonical);
+            assert(matrix.length!0 == matrix.length!1, "matrix must be square.");
         }
-        else
+        body
         {
-            if (!matrix.empty)
-            do
+            static if (kind == Contiguous)
             {
-                import mir.ndslice.algorithm: eachImpl;
-                eachImpl!fun(matrix.front!0, matrix.front!1);
-                matrix.popFront!1;
-                matrix.popFront!0;
+                import mir.ndslice.topology: canonical;
+
+                .eachUploPair!(fun, includeDiagonal)(matrix.canonical);
+            }
+            else
+            {
+                if (!matrix.empty)
+                {
+                    import mir.ndslice.algorithm: eachImpl;
+
+                    static if (includeDiagonal == true)
+                    {
+                        do
+                        {
+                            eachImpl!fun(matrix.front!0, matrix.front!1);
+                            matrix.popFront!1;
+                            matrix.popFront!0;
+                        }
+                        while (matrix.length);
+                    }
+                    else
+                    {
+                        do
+                        {
+                            eachImpl!fun(matrix.front!0[1 .. $],
+                                         matrix.front!1[1 .. $]);
+                            matrix.popFront!1;
+                            matrix.popFront!0;
+                        }
+                        while (matrix.length);
```

Comment:
`(matrix.length)` after `matrix.popFront!0;` is faster then `(matrix.length > 1)`